backend/src/main/java/uk/ac/cam/cl/kilo/data/MapMarker.java

```diff
@@ -0,0 +1,265 @@
+ *
+ * @author Nathan Corbyn
+ */
+public class MapMarker {
+  public static final String TABLE = "markers",
+      ID_FIELD = "id",
+      NAME_FIELD = "name",
+      DESC_FIELD = "desc",
+      X_FIELD = "x",
+      Y_FIELD = "y",
+      MAP_FIELD = "map_id";
+  private long id, map;
+  private String name, desc;
+  private int x, y;
+
+  /**
+   * Create a new map marker for the given map with the given name, description and position.
+   *
+   * @param map the map to mark
+   * @param name the name of the marker
+   * @param desc the description of the marker
+   * @param x the x coordinate of the marker
+   * @param y the y coordinate of the marker
+   * @throws IllegalArgumentException if the map is null or the name is null or empty
+   * @throws DatabaseException if the marker could not be added to the database
+   */
+  public MapMarker(ConferenceMap map, String name, String desc, int x, int y)
+      throws DatabaseException {
+    if (name == null || name.equals(""))
+      throw new IllegalArgumentException("Name must not be null or empty");
+    if (desc == null) desc = "";
+    if (map == null) throw new IllegalArgumentException("Map must bot be null");
+    this.name = name;
+    this.desc = desc;
+    this.map = map.getID();
+    this.x = x;
+    this.y = y;
+    try (Connection conc = Database.getInstance().getConnection()) {
+      PreparedStatement stmt =
+          conc.prepareStatement(
+              "INSERT INTO "
+                  + TABLE
+                  + "("
+                  + MAP_FIELD
+                  + ", "
+                  + NAME_FIELD
+                  + ", "
+                  + DESC_FIELD
+                  + ", "
+                  + X_FIELD
+                  + ", "
+                  + Y_FIELD
+                  + ") VALUES (?, ?, ?, ? , ?)");
+      stmt.setLong(1, this.map);
+      stmt.setString(2, name);
+      stmt.setString(3, desc);
+      stmt.setInt(4, x);
+      stmt.setInt(5, y);
+      stmt.executeUpdate();
+      ResultSet rs = stmt.getGeneratedKeys();
+      if (!rs.first()) throw new DatabaseException("Failed to generate ID for marker");
+      id = rs.getLong(ID_FIELD);
+    } catch (SQLException e) {
+      throw new DatabaseException(e);
+    }
+  }
+
+  private MapMarker(long id, long map, String name, String desc, int x, int y) {
+    assert (name != null);
+    assert (!name.equals(""));
+    assert (desc != null);
+    this.id = id;
+    this.map = map;
+    this.name = name;
+    this.desc = desc;
+    this.x = x;
+    this.y = y;
+  }
+
+  /** @return the ID of the marker */
+  public long getID() {
+    return id;
+  }
+
+  /** @return the name of the marker */
+  public String getName() {
+    return name;
+  }
+
+  /**
+   * @param name the new name for the marker
+   * @throws DatabaseException if the name could not be updated
+   * @throws IllegalArgumentException if the name is null or empty
+   */
+  public void setName(String name) throws DatabaseException {
```

Comment:
Were allowing multiple map markers with the same name? That complicates the events admin panel part as if events store the id of the location, a join is required to show the name, which I would like to show (so can we index on name instead?)

Reply:
The database connector won't enforce this because it is too expensive to check, if we want markers with unique names, SQL will. However, think of bathrooms etc. I think they should all be marked the same (i.e. 'toilet 1', 'toilet 2' sounds weird).

Reply:
In that case can we still do map markers by names and events will just assume that they are going on at all locations with the same name at once?

Reply:
Everything uses unique IDs (64 bit)